src/ngMessages/messages.js

```diff
@@ -330,63 +520,80 @@ angular.module('ngMessages', [])
     * ```html
     * <!-- using attribute directives -->
     * <ANY ng-messages="expression">
-    *   <ANY ng-message="keyValue1">...</ANY>
-    *   <ANY ng-message="keyValue2">...</ANY>
-    *   <ANY ng-message="keyValue3">...</ANY>
+    *   <ANY ng-message="expressionValue1">...</ANY>
+    *   <ANY ng-message="expressionValue2">...</ANY>
+    *   <ANY ng-message="expressionValue3">...</ANY>
     * </ANY>
     *
     * <!-- or by using element directives -->
     * <ng-messages for="expression">
-    *   <ng-message when="keyValue1">...</ng-message>
-    *   <ng-message when="keyValue2">...</ng-message>
-    *   <ng-message when="keyValue3">...</ng-message>
+    *   <ng-message when="expressionValue1">...</ng-message>
+    *   <ng-message when="expressionValue2">...</ng-message>
+    *   <ng-message when="expressionValue3">...</ng-message>
     * </ng-messages>
     * ```
     *
     * {@link module:ngMessages Click here} to learn more about `ngMessages` and `ngMessage`.
     *
-    * @param {string} ngMessage a string value corresponding to the message key.
+    * @param {expression} ngMessage|when an expression value corresponding to the message key.
     */
   .directive('ngMessage', ['$animate', function($animate) {
-    var COMMENT_NODE = 8;
     return {
-      require: '^ngMessages',
+      restrict: 'AE',
       transclude: 'element',
       terminal: true,
-      restrict: 'AE',
-      link: function($scope, $element, $attrs, ngMessages, $transclude) {
-        var index, element;
-
-        var commentNode = $element[0];
-        var parentNode = commentNode.parentNode;
-        for (var i = 0, j = 0; i < parentNode.childNodes.length; i++) {
-          var node = parentNode.childNodes[i];
-          if (node.nodeType == COMMENT_NODE && node.nodeValue.indexOf('ngMessage') >= 0) {
-            if (node === commentNode) {
-              index = j;
-              break;
-            }
-            j++;
-          }
-        }
+      require: '^ngMessages',
```

Comment:
Moreover it "should" be since it will then provide an error message if the user puts it in the wrong place.